### PR TITLE
[MAINT] refactor of tests and code in GLM module

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Relates to #3570

Changes proposed in this pull request:

- refactor and simplify complex tests in nilearn
- apply some sourcery suggested refactor in nilearn

### CODE REVIEW

File: nilearn/glm/tests/test_thresholding.py [hunk 1]
```diff
@@ -37,6 +33,19 @@ def test_fdr(rng):
...
+def test_fdr_error(rng):
+    n = 100
+    x = np.linspace(0.5 / n, 1.0 - 0.5 / n, n)
+    x[:10] = 0.0005
+    x = norm.isf(x)
+    rng.shuffle(x)
+
+    with pytest.raises(ValueError):
+        fdr_threshold(x, -0.1)
+    with pytest.raises(ValueError):
```
Comment (bthirion) [inline nilearn/glm/tests/test_thresholding.py:45]: Want to make the error explicit ?

**Review by bthirion — COMMENTED** [review]
LGTM overall

File: nilearn/glm/tests/test_thresholding.py (hunk 1 above)
Comment (Remi-Gau) [reply nilearn/glm/tests/test_thresholding.py:45]: good point will actually do that for all the "bare" `pytest.raises`

Comment (Remi-Gau) [reply nilearn/glm/tests/test_thresholding.py:45]: This makes me realize that bare pytest.raises were not catching we would expect from the comments in the test or from the logic of the test.

We should probably go through the test suite to ensure that error and warning messages are properly tested.

**Comment by jeromedockes** [comment]
> I am getting a failure on Windows only, on this test. Anyone can spot the issue here?

it seems the default encoding used on windows with a wester/american locale is more lenient than utf-8:

```python
import io

img_data = (
    b"GIF87a\x01\x00\x01\x00\xe7*\x00\x00\x00\x00\x01\x01"
    b"\x01\x02\x02\x07\x08\x08\x08\x0b\x0b\x0b\x0c\x0c\x0c\r;"
)
buf = io.BytesIO(img_data)
stream = io.TextIOWrapper(buf, encoding="cp1252")
stream.readline() # OK (but garbage output)
buf = io.BytesIO(img_data)
stream = io.TextIOWrapper(buf, encoding="utf-8")
stream.readline() # UnicodeDecodeError
```

btw,
- the documentation for `open` says a `ValueError` will be raised if an encoding error occurs (by default). UnicodeDecodeError is a subclass of ValueError so this is consistent with the behavior. but AFAICT there is no guarantee at least in the documentation that a UnicodeDecodeError will be used
- I'm not sure this check is useful, it basically raises the same error as the one that would propagate if it wasn't caught. actually it is less informative because the one raised by open specifies which encoding it tried, and that information is dropped by the nilearn rewording. also it is incorrect, because it states "does not appear to be a valid unicode text file", whereas it may be a perfectly well encoded text file, but encoded with a different encoding than the locale's default, which is used here

**Comment by jeromedockes** [comment]
also looking at the git blame for this file, it seems the black-formatting commits still show up in the git blame

**Comment by Remi-Gau** [comment]
> also looking at the git blame for this file, it seems the black-formatting commits still show up in the git blame

I know. 🙈 

Adding https://github.com/nilearn/nilearn/issues/3653 to the milestone for the next release.

**Comment by Remi-Gau** [comment]
> * I'm not sure this check is useful, it basically raises the same error as the one that would propagate if it wasn't caught. actually it is less informative because the one raised by open specifies which encoding it tried, and that information is dropped by the nilearn rewording. also it is incorrect, because it states "does not appear to be a valid unicode text file", whereas it may be a perfectly well encoded text file, but encoded with a different encoding than the locale's default, which is used here

Thanks for looking this up. I did not even where to look on this one.
So you would suggest nuking this test out of the test suite?

**Comment by jeromedockes** [comment]
> Thanks for looking this up. I did not even where to look on this one.

when you see different encoding errors or absence thereof on windows and unix, it helps to remember they don't use the same default encoding

another difference that comes up often in nilearn tests is windows refuses to unlink files that have an open file handle


> So you would suggest nuking this test out of the test suite?

the check in the code and the test

**Review by ymzayek — APPROVED** [review]
LGTM!

**Review by bthirion — APPROVED** [review]
LGTM, thx !

**Comment by Remi-Gau** [comment]
Will fix the error in the doc build in this PR: it is just a typo (apparently I cannot spell my own name).

File: doc/changes/latest.rst [hunk 2]
```diff
@@ -22,7 +22,7 @@ Enhancements
 - :bdg-primary:`Doc` Add backslash to homogenize :class:`~nilearn.regions.Parcellations` documentation (:gh:`4042` by `Nikhil Krish`_).
 - :bdg-success:`API` Allow passing Pandas Series of image filenames to :class:`~nilearn.glm.second_level.SecondLevelModel` (:gh:`4070` by `Rémi Gau`_).
 - Allow setting ``vmin`` in :func:`~nilearn.plotting.plot_glass_brain` and :func:`~nilearn.plotting.plot_stat_map` (:gh:`3993` by `Michelle Wang`_).
-- :bdg-success:`API` Support passing t and F contrasts to :func:`~nilearn.glm.contrasts.compute_contrast` that that have fewer columns than the number of estimated parameters. Remaining columns are padded with zero (:gh:`4067` by `Remi Gau`_).
+- :bdg-success:`API` Support passing t and F contrasts to :func:`~nilearn.glm.contrasts.compute_contrast` that that have fewer columns than the number of estimated parameters. Remaining columns are padded with zero (:gh:`4067` by `Rémi Gau`_).
```
Comment (Remi-Gau) [inline doc/changes/latest.rst:25]: @ymzayek 


Do you know why does the following throws a warning and fails in CI ?

```rst
:func:`~nilearn.glm.contrasts.compute_contrast`
```

Comment (ymzayek) [reply doc/changes/latest.rst:25]: oh I just opened a pr to fix it. It's how we have it set up in our doc/modules/glm.rst so compute_contrast is refrenced without the actual file name but just as a function of the module glm

**Pull request merged**